src/v/coproc/script_context.cc

```diff
@@ -177,31 +177,26 @@ ss::future<std::vector<process_batch_request::data>> script_context::read() {
       });
 }
 
-std::optional<storage::log_reader_config>
+storage::log_reader_config
 script_context::get_reader(const ss::lw_shared_ptr<ntp_context>& ntp_ctx) {
     auto found = ntp_ctx->offsets.find(_id);
     vassert(
       found != ntp_ctx->offsets.end(),
       "script_id must exist: {} for ntp: {}",
       _id,
       ntp_ctx->ntp());
-    const auto& cp_offsets = found->second;
-    if (cp_offsets.last_read != cp_offsets.last_acked) {
+    const ntp_context::offset_pair& cp_offsets = found->second;
+    const model::offset next_read
+      = (unlikely(cp_offsets.last_acked == model::offset{}))
+          ? model::offset(0)
+          : cp_offsets.last_acked + model::offset(1);
+    if (next_read <= cp_offsets.last_acked) {
         vlog(
           coproclog.info,
           "Replaying read on ntp: {} at offset: {}",
           ntp_ctx->ntp(),
           cp_offsets.last_read);
     }
-    const storage::offset_stats stats = ntp_ctx->log.offsets();
-    const model::offset next_read
-      = (unlikely(
-          cp_offsets.last_acked == model::model_limits<model::offset>::min()))
-          ? model::offset(0)
-          : cp_offsets.last_acked + model::offset(1);
-    if (next_read >= stats.committed_offset()) {
-        return std::nullopt;
```

Comment:
@mmaslankaprv Now that I've thought about this some more, we could always be in a situation where the next record or series of records is not of type raft::data_batch_type, which would mean the offsets we track for last read isn't exactly what the committed_offset() would be.

So I've made this method not return an optional and instead always return a reader, which will in turn perform a read and return an empty batch, for which I already have appropriate checks in because I am handling this situation of the reader returning an empty log when the record_batch doesn't match the filtered type (data_batch_type)

So my questions are:
1. Is there any performance considerations to doing things this way? I.e. Having the code path always create a reader and invoke a read, knowing it may return no data.
2. Would I ever be reading past the committed offset now?

Reply:
I think this is an intended way of doing reads, in general it is not often the case that you will have empty reader, you will never be reading past committed offset, I think you should read up to `high_watermark` tho

Reply:
ah unfortunately since copro doesn't depend on raft or cluster I don't think I can query this. What is the difference between reading up to `committed_offset` vs the `high_watermark`?

Reply:
@mmaslankaprv mentioned to me on slack we should be using the `dirty_offset` as `committed_offset` won't work well when acks= -1. However I think I remember Alex saying not the use the `dirty_offset` for some reason. @senior7515 can you chime in here?

Reply:
i can see a high-level difference in selecting an offset maintained by raft compared to one maintained at the storage layer. but considering only storage layer's offsets,  i'm not seeing much difference between dirty/stable(committed) offset. i could imagine something about durability / freshness but that could be controlled by a configuration option.

do you recall what you had in mind @senior7515 ?